### DISCUSSION TITLE
Structured Creativity: Exploring Metrics for AI-Generated Vector Art

### DISCUSSION THREAD

**Discussion by thorwhalen** [thorwhalen/essays#8] [General]
# [Structured Creativity: Exploring Metrics for AI-Generated Vector Art](https://medium.com/@thorwhalen1/structured-creativity-exploring-metrics-for-ai-generated-vector-art-b7b182f6b68b?source=friends_link&sk=66f0aa5a19178074c3649ac2f13068a9)

Large Language Models (LLMs) are trained on vast datasets, guided by internal metrics focused on things like token prediction accuracy. While these metrics are crucial for the model's overall development, they often fall short when it comes to evaluating how well an AI performs a specific task for a user. We, as users of AI 'agents,' need our own evaluation frameworks, tailored to our unique objectives and sensitive to the specific nuances of our work. The shift is clear: we've moved from being producers to curators, placing a premium on clearly defining our goals and developing metrics to measure AI performance against them. This is no longer a luxury, but a necessity, allowing us to select the right tools, fine-tune their output, and ultimately, make informed decisions about AI integration. 

The challenge, of course, lies in quantifying these objectives. 
AI evaluation methods span multiple dimensions—[objective vs. subjective](https://arxiv.org/abs/2009.08593), [closed-set vs. open-ended](https://arxiv.org/abs/2009.08593), [reference-based vs. reference-free](https://arxiv.org/abs/2009.08593), and [intrinsic vs. extrinsic](https://arxiv.org/abs/2009.08593)—each capturing distinct aspects of model performance.

Some tasks, like software development, are mercifully straightforward. 'Does it compile? Does it crash? Does it give me the right answer?' Easy peasy. 
We find ourselves, if not exactly blessed, then certainly less burdened by the existential dread of evaluative subjectivity.
Then you get to image generation, and suddenly it's 'Is it pretty?' And we're like, 'Dude, what's pretty?' A cruel joke played on our finite cognitive capacities. We're screwed. 

As a much humbler investigation I wanted to see how well different AIs would do in SVG generation tasks whose performance I can measure easily: Can they make an image of a word with some specific constraints. What's appealing about this image generation evaluation is that:
* it generates SVG code, which could not at all compile (a first, objective step of the evaluation)
* we can (for some types of constraints) verify, objectively, in the code, whether these constraints have been adhered to.
* we can OCR the image to see if it resolves to the input word

This has all the characteristics of the kind of evaluation we strive for!

I will post some python code for this kind of evaluation in a different article (to not scare the no-code readers away).
In the second part of this article is one fun POC evaluation of how different AIs fared on this task, for one particular word and constraint. Actually, the whole point of this post, originally was to just slap that up there for my own reference. Then I felt it needed an introduction, and then this happened. 

And I'll digress just a little longer on this question: Why SVG generation?

## Beyond Pixels: Evaluating Video Generation and Transitioning to Vectorized Representations

For some time now, I've been trying to keep an eye on how the "structural" (or "geometric") approaches of AI are going.
These methods exist, but seem to be quite niche next to the "point-based" methods that have been so successful for that last 10+ years now. (Look, I could go on for days about this, seriously, pages and pages. But this is supposed to be a simple intro, to something actually quite pedestrian, not a PhD thesis, so let's keep it short, okay?) As far as image and video generation goes, these structural methods would be things like, just to mention two:
* Vector-based generation: Taking simple shapes as the building blocks (see [Adobe Illustrator text to vector AI Features](https://www.adobe.com/products/illustrator/text-to-vector-graphic.html) or [Recraft Blog](https://www.recraft.ai/blog/best-ai-vector-generators)), or
* Scene representation generation, which, for instance, take objects, lighting, and physical laws as the building blocks ([NVIDIA Technical Blog](https://developer.nvidia.com/blog/rapidly-generate-3d-assets-for-virtual-worlds-with-generative-ai/) and [Google AI Blog](https://ai.googleblog.com/2022/10/dreamfusion-text-to-3d-using-2d-diffusion.html, [Scene Representation Networks](https://arxiv.org/html/2402.00028v1)). 

Recent advances in video generation have demonstrated that state‐of‐the‐art models can create visually impressive and temporally coherent videos from text prompts. At a low level, objective metrics such as the Fréchet Video Distance (FVD), Inception Score (IS), and CLIP-based similarity have long been used to guide neural networks toward producing realistic video content. These metrics evaluate aspects like pixel distribution, frame consistency, and overall visual fidelity. However, they often fall short in assessing how accurately a generated video aligns with the user's intent—a gap that has spurred research into metrics that account for text–video alignment and motion dynamics. For instance, recent work on [T2VScore](https://arxiv.org/abs/2401.07781) has sought to integrate text–video alignment and quality assessment, while frameworks like [EvalCrafter](https://arxiv.org/abs/2310.11440v3)provide a more comprehensive evaluation by considering motion quality and temporal consistency.

Another promising direction involves reducing the output space from millions of pixels to a more compact, semantically meaningful representation. By encoding videos (or images) in a vectorized form—such as diagrams or cartoon-like scenes expressed as Scalable Vector Graphics (SVG)—the parametric space is dramatically reduced. This offers several potential advantages:

- **Simpler, interpretable representations:** Vectorized outputs, unlike raw pixel matrices, consist of a series of high-level drawing commands (e.g., lines, curves) that explicitly capture structural information. Such representations are inherently resolution-independent and easier to manipulate or edit.
- **Enhanced semantic control:** With a reduced parametric space, the generated content can be more directly tied to user prompts. For example, a cartoon-like video composed of vectorized shapes might be evaluated not just on its low-level visual quality but also on its adherence to stylistic and structural guidelines derived from the text prompt.
- **Facilitated evaluation of user intent:** Evaluating videos in a vectorized domain can simplify the development of metrics that directly assess whether the generated video fits the intended narrative or design. As world models begin to integrate higher-level semantic information, the gap between user expectations and objective metrics may narrow.

Recent projects such as [NeuralSVG](https://webdesignerdepot.com/neuralsvg-revolutionizing-vector-graphics-with-ai/) and **SVGFusion** exemplify this trend. NeuralSVG uses neural networks to generate editable SVG images from text prompts, effectively capturing design logic in a compact form. Although primarily focused on static graphics, the underlying principles—reducing the complexity of representation and enabling fine-grained control—can be extended to video generation. In the near future, we may see AI tools that produce cartoon-like or diagrammatic videos where every element is represented by vector primitives, allowing for more transparent and semantically meaningful evaluation metrics.

By transitioning from traditional pixel-based video generation to approaches that leverage vectorized representations, the field could achieve not only more efficient generation pipelines but also more interpretable and controllable outputs that align closely with user intent.

# Exploring AI-Generated SVGs: A Proof-of-Concept

Now what follows, is all I wanted to post originally. Just to have it readily available to illustrate some differences between various AIs. 
This is a series of informal, anecdotal tests in which various AI systems were asked to generate an SVG image for a very specific task. 
The prompt was:

> *"Give me an SVG image of the word 'figiri' using a minimal set of identical shapes (same dimensions, etc., but you are allowed to rotate them)"*

After receiving a generated image or code, I followed up with requests like:

- *"Can you give me SVG code for that?"*
- *"Can you render this SVG as an image for me?"*

For each system, I copied the resulting SVG code into [Boxy SVG](https://boxy-svg.com/app) to verify the rendered output. **Note:** This is not a formal evaluation system. It’s just a fun proof-of-concept—an initial exploration into how one might develop an evaluation framework for SVG generation. A proper evaluation would involve running many such tests over various prompts and using objective metrics (such as feature-vector analysis or even AI-based evaluators) to measure how well the generated image meets the intended description.

Below are my anecdotal results from several AI systems.

---

### Gemini (2.0 Flash)
The first attempt with Gemini (using the "Flash" variant) was, in a word, *so bad!* It generated SVG code that, when rendered, produced the following image:

<img width="403" alt="Gemini 2.0 Flash SVG output" src="https://github.com/user-attachments/assets/e9cbe7a1-c475-4ed8-839a-b898110daffc" />

When I asked Gemini to render it, it replied:  
*"I'm a text-based AI and can't assist with that."*  
However, when I pasted the code into Boxy SVG, this is what I saw.

### Gemini (2.0 Flash Thinking Experimental)
A slightly different version—dubbed "Flash Thinking Experimental"—produced a marginally better result where you can almost make out the word "figiri":

<img width="280" alt="Gemini 2.0 Flash Thinking Experimental SVG output" src="https://github.com/user-attachments/assets/54ba47bc-888a-4c67-ad0e-ea3b2529fa0e" />

---

### ChatGPT "4o"
Using ChatGPT (specifically GPT-4o), the initial response was a bit confused about the word "figiri" (asking, "Figiri? In what language/script?"), but eventually produced an SVG along with an image preview:

<img width="313" alt="ChatGPT GPT-4o initial preview" src="https://github.com/user-attachments/assets/f50da64b-afa3-44ba-b23b-13576e590dee" />

After further prompting, ChatGPT rendered the SVG image almost identically:

<img width="499" alt="ChatGPT rendered SVG output" src="https://github.com/user-attachments/assets/8cce2265-a6ee-4dc4-b780-c3302c3e7158" />

---

## 03-mini-high

This system performed even worse. The output was unsatisfactory:

<img width="253" alt="03-mini-high SVG output" src="https://github.com/user-attachments/assets/b65e273e-d7c9-4d03-8930-256bedf03667" />

When I requested it to render the SVG as an image, it produced SVG code that, when rendered (using Boxy SVG), resulted in an image like this:

![SVG "figiri" Image](data:image/svg+xml;utf8,<svg%20xmlns=%22http://www.w3.org/2000/svg%22%20width=%22600%22%20height=%22100%22%20viewBox=%220%200%20600%20100%22><defs><symbol%20id=%22bar%22%20viewBox=%22-15%20-5%2030%2010%22><rect%20x=%22-15%22%20y=%22-5%22%20width=%2230%22%20height=%2210%22%20fill=%22black%22/></symbol></defs><g%20transform=%22translate(0,0)%22><use%20xlink:href=%22#bar%22%20transform=%22translate(20,50)%20rotate(90)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(35,35)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(35,50)%22/></g><g%20transform=%22translate(100,0)%22><use%20xlink:href=%22#bar%22%20transform=%22translate(50,50)%20rotate(90)%22/></g><g%20transform=%22translate(200,0)%22><use%20xlink:href=%22#bar%22%20transform=%22translate(20,50)%20rotate(90)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(50,20)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(50,80)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(80,65)%20rotate(90)%22/></g><g%20transform=%22translate(300,0)%22><use%20xlink:href=%22#bar%22%20transform=%22translate(50,50)%20rotate(90)%22/></g><g%20transform=%22translate(400,0)%22><use%20xlink:href=%22#bar%22%20transform=%22translate(20,50)%20rotate(90)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(35,35)%22/><use%20xlink:href=%22#bar%22%20transform=%22translate(40,65)%20rotate(45)%22/></g><g%20transform=%22translate(500,0)%22><use%20xlink:href=%22#bar%22%20transform=%22translate(50,50)%20rotate(90)%22/></g></svg>)

---

## Grok 3

The Grok 3 system delivered a range of outputs. They were definitely cool. But… They were just images. Not SVG-based. Here's a bunch of them:

![grok_figiris](https://github.com/user-attachments/assets/0e2d244a-72e1-412d-9afa-a953bdceb093)

And when I asked it to make some equivalent SVG code, well, yeah, not great:

<img width="271" alt="Grok 3 rendered SVG output" src="https://github.com/user-attachments/assets/8bb43074-4b66-4c48-9db2-eac2e93f0053" />

---

## Claude (3.7 Sonnet)

Claude (version 3.7 Sonnet) was the clear winner in these tests. Although its output isn’t particularly beautiful, it was correct and reproducible. Claude provided its answer in a dual-tab window—one showing a live preview (SVG rendered as an image) and another displaying the SVG code. When I requested the code, Claude’s output rendered exactly as shown:

<img width="703" alt="Claude 3.7 Sonnet SVG output" src="https://github.com/user-attachments/assets/2f07815e-d11d-4453-a20a-dc6364cb1f35" />

---

## Final Thoughts

These anecdotal tests are not a final evaluation, but rather an initial exploration into how different AI systems handle the task of generating structured images—in this case, SVGs. They provide a glimpse into the broader ambition of developing robust evaluation methods that can assess AI's ability to produce structured images and videos. By experimenting with a specific prompt and examining both the generated code and its rendered output, we can begin to understand what objective metrics might look like for this kind of task. 

This proof-of-concept lays the groundwork for a future evaluation framework where multiple prompts are tested and quantified using objective measures—such as feature-vector comparisons or AI-driven assessments of how well the generated output aligns with the intended description. Such an approach would not only advance our understanding of AI's generative capabilities in the vectorized domain, but also contribute to the larger effort of crafting task-specific evaluation methodologies. These methods could, in turn, be applied to other domains where structured outputs are critical, ensuring that we have reliable, objective metrics to gauge performance across diverse AI applications.

The journey toward comprehensive evaluation is just beginning, and these experiments are an important stepping stone in that ongoing process.

**Comment by thorwhalen** [comment]
## Python code for SVG generation evaluator

```py
# builtins
import io, operator
# stuff that might need installing, if not already installed
import cairosvg, pytesseract, PIL, dol  # pip install cairosvg, pytesseract, pillow, dol

Img = PIL.Image.Image

def svg_to_img(svg: str) -> Img:
    """Convert SVG string to PIL Image."""
    png_data = cairosvg.svg2png(bytestring=svg.encode('utf-8'))
    img = Image.open(io.BytesIO(png_data))

    return img


def img_to_text(img: Img) -> str:
    """Extract text from image using OCR."""
    text = pytesseract.image_to_string(img)
    return text.strip()


svg_to_text = dol.Pipe(svg_to_img, img_to_text)


def text_to_svg(text: str, font_size: int = 40) -> str:
    """Convert text to SVG."""
    return f'''
    <svg xmlns="http://www.w3.org/2000/svg" width="300" height="100">
        <rect width="100%" height="100%" fill="white"/>
        <text x="10" y="50" font-family="Arial" font-size="{font_size}" fill="black">{text}</text>
    </svg>
    '''


def text_to_svg_score_for_text(
    text='test text',
    text_to_svg=text_to_svg,
    *,
    svg_to_text=svg_to_text,
    comparison_func=operator.eq,
):
    """Test the text-to-SVG conversion."""
    svg = text_to_svg(text)
    img = svg_to_img(svg)
    ocr_text = img_to_text(img)

    return comparison_func(ocr_text, text)


# Test with different words
test_text = ["Hello", "Python", "OCR", "Computer Vision"]
for text in test_text:
    score = text_to_svg_score_for_text(text)
    # print with justification
    print(f"Text: {text:<20} Score: {score}")
```
